### PR TITLE
feat(integration): Add KubeflowCallback to enable automatic progress …

### PR DESCRIPTION
**Pull request by abhijeet-dhumal**
# What does this PR do?

Fixes #44486

Adds `KubeflowCallback` to enable automatic progress and metrics reporting for training jobs running on [Kubeflow Trainer](https://github.com/kubeflow/trainer).

When training runs inside a Kubeflow TrainJob, the callback automatically:
- Reports training progress percentage (0-100%)
- Calculates and reports estimated time remaining (ETA)
- Pushes training metrics (loss, accuracy, etc.) to the TrainJob status

This enables real-time visibility into training progress via standard Kubernetes APIs:

```bash
kubectl get trainjob my-job -o jsonpath='{.status.trainerStatus}'
# {"progressPercentage": 67, "estimatedRemainingSeconds": 3600, "metrics": [...]}
```

**Zero friction for users** - no code changes required:

```python
from transformers import Trainer, TrainingArguments

trainer = Trainer(model=model, args=TrainingArguments(...), train_dataset=ds)
trainer.train()  # Progress automatically reported when running in Kubeflow
```

## Changes

- Add `is_kubeflow_available()` function that detects Kubeflow environment via `KUBEFLOW_TRAINER_STATUS_URL` env var
- Add `KubeflowCallback` class following the `MLflowCallback`/`WandbCallback` pattern
- Register in `INTEGRATION_TO_CALLBACK` and `get_available_reporting_integrations()`
- Add unit tests in `tests/trainer/test_trainer_callback.py`

## Related PRs

| Repository | PR | Description |
|------------|-----|-------------|
| kubeflow/trainer | [#3227](https://github.com/kubeflow/trainer/pull/3227) | Controller-side implementation (status server) |
| kubeflow/sdk | https://github.com/kubeflow/sdk/issues/367 | SDK `update_runtime_status()` utility |

---

# Additional context

This is part of [KEP-2779](https://github.com/kubeflow/trainer/issues/2779) - a coordinated effort with the Kubeflow community to enable real-time training progress tracking on Kubernetes.

The callback has an optional dependency on kubeflow-sdk. The SDK handles:
- HTTP transport to the Kubeflow controller
- Authentication via projected service account tokens
- Throttling (max 1 update per 5 seconds)

When the `TrainJobProgress` feature gate is enabled in Kubeflow Trainer, the controller automatically injects these environment variables into training pods:
- `KUBEFLOW_TRAINER_SERVER_URL` - HTTPS endpoint for status updates
- `KUBEFLOW_TRAINER_SERVER_CA_CERT` - CA cert for TLS
- `KUBEFLOW_TRAINER_SERVER_TOKEN` - Projected SA token for auth

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by abhijeet-dhumal** [comment]
Waiting for Kubeflow Trainer feature implementation to be merged : https://github.com/kubeflow/trainer/pull/3227 ...

File: src/transformers/training_args.py [hunk 1]
```diff
@@ -1525,6 +1525,12 @@ def __post_init__(self):
         elif not isinstance(self.report_to, list):
             self.report_to = [self.report_to]
 
+        # Auto-enable Kubeflow integration when running inside a Kubeflow TrainJob
+        from .integrations import is_kubeflow_available
+
+        if is_kubeflow_available() and "kubeflow" not in self.report_to:
+            self.report_to = list(self.report_to) + ["kubeflow"]
+
```
Comment (SunMarc) [inline src/transformers/training_args.py:1533]: Can we update `get_available_reporting_integrations` so that it works with report_to == "all". Also, instead of enabling it here by default, can we force the user to pass `report_to="kubeflow"` in kubeflow instead or it is a bit complicated ?

**Review by SunMarc — APPROVED** [review]
Thanks for the PR ! Just a small question

**Comment by SunMarc** [comment]
Let me know when we can merge this

File: src/transformers/integrations/integration_utils.py [hunk 2]
```diff
@@ -2354,6 +2362,192 @@ def on_predict(self, args, state, control, metrics, **kwargs):
...
+
+    This callback is automatically registered when training inside a Kubeflow TrainJob with the
+    `TrainJobRuntimeStatus` feature gate enabled. The Kubeflow controller injects the required
+    environment variables into the training pod.
+
+    **Environment Variables (injected by controller):**
+
+    - `KUBEFLOW_TRAINER_STATUS_URL`: HTTPS endpoint for status updates
+    - `KUBEFLOW_TRAINER_STATUS_CA_CERT`: Path to CA certificate for TLS verification
+    - `KUBEFLOW_TRAINER_STATUS_TOKEN`: Path to service account token for authentication
```
Comment (astefanutti) [inline src/transformers/integrations/integration_utils.py]:
```suggestion
    - `KUBEFLOW_TRAINER_SERVER_URL`: HTTPS endpoint for status updates
    - `KUBEFLOW_TRAINER_SERVER_CA_CERT`: Path to CA certificate for TLS verification
    - `KUBEFLOW_TRAINER_SERVER_TOKEN`: Path to service account token for authentication
```

File: src/transformers/training_args.py (hunk 1 above)
Comment (abhijeet-dhumal) [reply src/transformers/training_args.py:1533]: Thanks for the feedback @SunMarc! 

I've removed the auto-enable block from `training_args.py`, Kubeflow is now included in `get_available_reporting_integrations()` when the env var is detected. 
This means `report_to="all"` (the default) automatically includes kubeflow when running inside a TrainJob.. Users who explicitly set `report_to` to something else won't have kubeflow added automatically

This follows the same pattern as other integrations (wandb, mlflow, etc.) - available when detected, included by default

File: src/transformers/integrations/integration_utils.py (hunk 2 above)
Comment (abhijeet-dhumal) [reply src/transformers/integrations/integration_utils.py]: Thanks @astefanutti, I have updated these env vars accordingly ✅

File: src/transformers/training_args.py (hunk 1 above)
Comment (SunMarc) [reply src/transformers/training_args.py:1533]: Oh, note that the default is not "all" anymore from v5 onwards. This might complicate the usage in kubeflow i think.

Comment (abhijeet-dhumal) [reply src/transformers/training_args.py:1533]: Thanks for flagging this @SunMarc! With report_to defaulting to "none" in v5+, users would need to explicitly set report_to="kubeflow" to enable progress reporting.
So in that case, do you think re-adding auto-enable make sense? - If Kubeflow env detected, add to report_to regardless of setting?

Comment (SunMarc) [reply src/transformers/training_args.py:1533]: Well, I think it is fine to reintroduce it. Since the check if not just having the lib installed but instead a global var needs to be set, I feel like this is okay.

Comment (abhijeet-dhumal) [reply src/transformers/training_args.py:1533]: Done, ready for review 🏁

**Review by SunMarc — APPROVED** [review]
Thanks for iterating !
